### ISSUE TITLE
✨ Necessary about:config protect privacy againtst fingerprint

### ISSUE THREAD

**Issue by xenonuuuy** [privacytoolsIO/privacytools.io#878]
hi guys, 

we should add this line to about:config manually : **font.system.whitelist**
right click and string then paste this value : 

_Arial, Arial Black, Batang, 바탕, Cambria Math, Courier New, Euphemia, Gautami, Georgia, Gulim, 굴림, GulimChe, 굴림체, Iskoola Pota, Kalinga, Kartika, Latha, Lucida Console, MS Gothic, ＭＳ ゴシック, MS Mincho, ＭＳ 明朝, MS PGothic, ＭＳ Ｐゴシック, MS PMincho, ＭＳ Ｐ明朝, MV Boli, Malgun Gothic, Mangal, Meiryo, Meiryo UI, Microsoft Himalaya, Microsoft JhengHei, Microsoft JengHei UI, Microsoft YaHei, 微软雅黑, Microsoft YaHei UI, MingLiU, 細明體, Noto Sans Buginese, Noto Sans Khmer, Noto Sans Lao, Noto Sans Myanmar, Noto Sans Yi, Nyala, PMingLiU, 新細明體, Plantagenet Cherokee, Raavi, Segoe UI, Segoe UI Emoji, Shruti, SimSun, 宋体, Sylfaen, Tahoma, Times New Roman, Tunga, Verdana, Vrinda, Yu Gothic UI_

tor browser have this about:config but i have add new font  ( Segoe UI Emoji ) to show emoji .

before enter this value you can see your fingerprint on[ this page ](https://panopticlick.eff.org)( with resist fingerprint protection it is **uniqe fingerprint** still but when add this value you see fingerprint change to  **partial protection**

**Comment by blacklight447-ptio** [comment]
I would be okay with this change

**Comment by beerisgood** [comment]
@Thorin-Oakenpants 
What did you think about?

**Comment by Thorin-Oakenpants** [comment]
No. I'll explain further tomorrow. But for starters, that's the windows whitelist for Tor Browser, but you're NOT tor browser, so you will become super unique. Your best best is to block document fonts.

Also, the whitelist has a typo: `Microsoft JengHei UI` is misspelled (Microsoft J**h**engHei UI`), it's also missing the bundled Mozilla emoji font, not to mention Tor Browser's own different bundled emoji font. And why are you whitelisting Noto Sans fonts that do not exist on your system (these are ones bundled by TB). The panopticlick test is giving you a false reading, because it's not checking much. Even the tor browser leaks entropy across windows fonts (I know of at least 7 results, and suspect dozens more) - so stop kidding yourself, and just block document fonts

And, please stop using entropy figures from these test sites: they are tainted

**Comment by blacklight447-ptio** [comment]
> No. I'll explain further tomorrow. But for starters, that's the windows whitelist for Tor Browser, but you're NOT tor browser, so you will become super unique. Your best best is to block document fonts.
> 
> Also, the whitelist has a typo: `Microsoft JengHei UI` is misspelled (Microsoft J**h**engHei UI`), it's also missing the bundled Mozilla emoji font, not to mention Tor Browser's own different bundled emoji font. And why are you whitelisting Noto Sans fonts that do not exist on your system (these are ones bundled by TB). The panopticlick test is giving you a false reading, because it's not checking much. Even the tor browser leaks entropy across windows fonts (I know of at least 7 results, and suspect dozens more) - so stop kidding yourself, and just block document fonts
> 
> And, please stop using entropy figures from these test sites: they are tainted

so in short: we should block document fonts?

**Comment by Thorin-Oakenpants** [comment]
> but i have add new font ( Segoe UI Emoji ) to show emoji

The font for Mozilla is `Twemoji Mozilla` and is bundled with your application. Not everyone will have `Segoe UI Emoji`

**Comment by Thorin-Oakenpants** [comment]
> so in short: we should block document fonts?

disclosure: I am not a font expert, just a bit knowledgeable on what the browser can report etc. And I'm an English user, so I have no real insight on the impact for non-English users.

First of all, controlling 3rd party scripts will solve a lot of FP'ing. But lets ignore that (and some scripts like reddit are 1st party and becoming more prevalent) and assume the worst: i.e the script is running. Always assume the worst.

Setting `browser.display.use_document_fonts` -> `0`: This is in your options under General > Language and Appearance > Fonts & Colors > Advanced > Allow pages to choose...

- This blocks the most common font fingerprinting technique used which I call **fpjs2**
- Sites want to use scripts that don't impact the site's performance and fpjs2 type FPing is lightweight. - The perf impact is directly related to how many fonts are tested
- Not many fonts are needed to be tested to quickly raise your entropy, and testing more after that quickly drops off in efficiency
- If they really wanted to (and I have done some work on this), they could run smaller crafted font lists based on your OS (which is trivial to know), and thus be more effective (i.e still check x amount of fonts, just be smarter about it -> higher entropy): e.g why test for Mac fonts on a windows system, instead test for font entropy within that OS: for example, on Windows, the browser will report some fonts differently depending on whether it uses GDI or DirectWrite (something about how fonts are decided as part of a font family). So some FF browsers will report Arial Narrow and some won't, despite all having it. And of course, you would want to test for some super common fonts provided by software, such as adobe, open/libre office etc.
- It's not hard to run a small super fast list that gets maximum entropy

In this most common font FPing technique: **if the document can't set a font, then it can't be detected**. The panopticlick test is a bit naff: if always detects two fonts, even if the system doesn't have them (wingdings and something else from memory: although this affects everyone, so no big deal). It's a short list aimed at those who don't block doc fonts. It should return zero fonts (their list), but in reality, in Windows, there are at least 18 fonts you **cannot** block. Set your whitelist to a single non-existent font and restart. I can still detect up to 18 fonts on your system. I'm not sure what happens on other systems (but I'm working on it).

Another thing to consider, is how broken the web can become when you block document fonts, for non English users. So this solution certainly doesn't work for all. Obviously your language settings have default fonts: so those fonts are not blocked: e.g `Armenian` will allow `Sylfaen`, or `Simplified Chinese` -> SimSun. But I have no idea what each language does to the end-user experience. Maybe suddenly all Latin chars turn into `tofu` (the empty box or other marker shown in place of an undisplayable code point)

I can get font version entropy due to changes between versions (I can test for added unicode codepoints) - eg Times New Roman and Arial - blocking doc fonts means I can never be sure I am testing the right font (only the default).

I can get css line height => entropy: but not if document fonts are blocked: well, I can still get the value/some-entropy: but it would be based on your default font.

So blocking document fonts does a lot more than stop font list enumeration, it can also affect other FP'ing tests. But it's about the only viable solution for Firefox users right now, because it's a single option in the UI, and because en-US users are a very large set.

--

The **only** solution (as a start) is to bundle all fonts, the same fonts and same font versions for all users (per OS is fine, since it's trivial to detect your platform), and whitelist these fonts - but that requires bandwidth traffic and will not happen any time soon. And it also probably needs something done to handled fonts rather than font families? And this solution only works if it is enforced in a set of users. Some variance may be acceptable: e.g the Arial Narrow on Windows may cause two buckets of windows users, but these are large groups and a solution may not be feasible.

PS: On Tor Browser (in my windows 7), I can detect `67` fonts. If I block document fonts in Tor Browser (as an experiment: don't keep this change permanently: also TB uses some extra patches, so this is different to Firefox), I detect `zero`. But by using a different (slower) technique, I can actually still detect 49 of them. But this technique requires downloading a font: so if you also have to think about controlling that (and web breakage).

There is no easy answer. Whitelisting will make you super unique: it doesn't matter that you reduced the number of fonts found, you'll likely be almost the only one returning that list: because it has to be applied manually (unless you make your user.js `UTF-8` compliant or something), and almost no-one in the world does this. But if you block document fonts (and can live with any breakage), which has a single pref and even a UI option .. this will be magnitudes of order more common (and would block 99% of font FPs out there, IMO).

Wow, wrote far too much, and its a terrible piece of writing, all disjointed and what-not. Apologies.

tl;dr: nah, just use document fonts pref for en-US users with a warning and never tell people to whitelist fonts (leave that for RFP to implement properly)

**Comment by Mikaela** [comment]
Blocking document fonts can have unforeseen consequences even if the pages were just in English. From experience Bitwarden's symbols became invisible and Pleroma's reply and boost and similar buttons became boxes.

I don't remember which font I enforced.

**Comment by beerisgood** [comment]
@Mikaela every blocked stuff can make trouble. It's always comfort vs security/ privacy

**Comment by Thorin-Oakenpants** [comment]
> become boxes

Because the font specified (doc font) is pointing at a PUA that your (default) font doesn't have. That said, most glyphs (icon fonts) these days are downloadable fonts (any decent web dev would not rely on a system font). There is also the case that web extensions should be exempt from some "things": such as some RFP measures. Here's a recent example of a fix for web extensions: https://bugzilla.mozilla.org/show_bug.cgi?id=1564208

**Comment by blacklight447-ptio** [comment]
After some thinking, there are to many issues that could up when going thru with this change, while the privacy winning aspect seems to be minimal, so im closing the issue, feel free to comment with new arguments to open it up again.
